### PR TITLE
add dynamics from forces solver and ignition

### PR DESCRIPTION
**Pull request by nicolaloi**
taken from gokart-pympc

### CODE REVIEW

File: waymax/utils/gokart_config.py [hunk 1]
```diff
@@ -7,20 +7,26 @@
...
     l1: float = 0.72  # Distance from cog to front tires
     l2: float = 0.47  # Distance from cog to rear tires
     w1: float = 0.94  # Distance between front Tires
     w2: float = 1.08  # Distance between rear Tires
+    h: float = 0.24   # Height of the cog
     back2backaxle: float = 0.23  # Distance from the rear of the gokart to the back axle
     frontaxle2front: float = 0.33  # Distance from the front axle to the front of the kart
     wheel2border: float = 0.18  # Side distance between center of the wheel and external frame
     F2n: float = l1 / (l1 + l2)  # Normal force at the rear axle "portion of Mass supported by rear tire"
+    m: float = 335  # Mass kg of the gokart
```
Comment (alezana) [inline waymax/utils/gokart_config.py]: with someone in it?

File: waymax/utils/gokart_config.py [hunk 2]
```diff
@@ -7,20 +7,26 @@
...
+    
+    def __post_init__(self):
+        assert self.l == self.l1 + self.l2, f"l {self.l} must be equal to l1 {self.l1} + l2 {self.l2}"
```
Comment (alezana) [inline waymax/utils/gokart_config.py]: compute directly self.l here,  use fiel init False in dataclass field

File: waymax/dynamics/tricycle_model.py [hunk 3]
```diff
@@ -73,9 +74,10 @@ def __init__(
     self._max_steering = max_steering
     self._normalize_actions = normalize_actions
 
-  def action_spec(self) -> specs.BoundedArray:
+  def action_spec(self, opposite_normalize_actions = False) -> specs.BoundedArray:
```
Comment (alezana) [inline waymax/dynamics/tricycle_model.py]: missing docs, what is meant by "opposite" ?

File: waymax/utils/gokart_config.py [hunk 4]
```diff
@@ -7,20 +7,26 @@
...
 
 
 @dataclass
 class TricycleParams:
     Iz: float = 0.7  # Inertia around the z axis
-    REG_: float = 0.5  # Regularization factor
+    REG_: float = 0.1  # Regularization factor
```
Comment (alezana) [inline waymax/utils/gokart_config.py]: regularization of what? more docs please

**Review by alezana — CHANGES_REQUESTED** [review]

**Comment by alezana** [comment]
I see some possible simplifications. In particular I don't understand the introduction of the `normalize_actions : Optional[bool] = None`, it makes the code heavier and I don't see why it's needed.
I would do the following:
- pass the bool on normalized or not in the __init__,
- based on that value it defines the returned vector from specs.Action.
- then once you get some actions in, you clip the values according to specs.Action and rescale to normal values only if you are using normalized actions.

I will push soon this, I also moved max steering, acc to teh parameters of the model. not sure why they were kept separate

**Comment by alezana** [comment]
@nicolaloi I implemented the things above but please double check correctness.
Also please implement some short unit tests e.g. step and verify orientations, velocities etc.
Such that we are more confident in the future that signs, positions etc don't get mistaken

**Review by alezana — CHANGES_REQUESTED** [review]
see comments above ^

**Comment by nicolaloi** [comment]
add option to choose among 1) original, 2) forces, or 3) ignition dynamics. Trajectory now has extras terms (acc_x, acc_y) needed by the ignition dynamics.

**Comment by alezana** [comment]
@nicolaloi did you test learning with these models? shall I merge?

**Comment by nicolaloi** [comment]
yes 

https://github.com/user-attachments/assets/1a2fa5eb-3df7-4003-bb86-fdb55c0d9452

**Pull request merged**